### PR TITLE
cmd/exec: adds --stdin-input (-I) flag for input piping or manual entry

### LINKED ISSUE #6538: Add --stdin-input flag to opa exec command
## What is the underlying problem you're trying to solve?
I am trying to figure out how to execute OPA inside a standard AWS Lambda. The requirements are:
- Configure OPA through Discovery API;
- Collect Status and Decision Logs telemetry;
- Use any kind of built-in function (SDK dependent or not);
- Take into account that a Lambda is not alive all the time.

So I'm working in a POC to see if the `opa exec` command is a good fit, but in my tests, I saw that the `opa exec` command executes against input files. I can write an input.json file on the filesystem before invoking the command, but it would be cheaper if it were possible to pass input data from stdin, so I could dynamically build my input together command and invoke OPA.

For example:

`echo my_input | opa exec --stdin-input ...`

## Describe the ideal solution

Adding the new flag, as we have on the `opa eval` command, will make the use of OPA more flexible when using OPA as a command inside scripts.

## Describe a "Good Enough" solution

Maybe creating an input.json before invoking the OPA command, but this solution is not so efficient and more expensive.

## Additional Context

There is an OPA integration with AWS API Gateway, but in my case, I'm trying to enforce authorization policies inside a standard AWS Lambda that could be written in any language. WebAssembly will not work because I need to use any kind of built-in function from Rego code.

### PR DESCRIPTION
**Pull request by colinjlacy**
### What are the changes in this PR?

- adds `--std-in` flag to `exec` command
- adds test coverage to `cmd/internal/exec` package
- updates `exec` command docs with new flag

### Notes to assist PR review:

Can test with input piping, e.g. `echo "{\"foo\": 8}" | opa exec ...` or manual entry via Stdin.

Fixes #6538

### CODE REVIEW

File: cmd/exec.go [hunk 1]
```diff
@@ -78,6 +79,7 @@ e.g., opa exec --decision /foo/bar/baz ...`,
 	cmd.Flags().VarP(params.LogLevel, "log-level", "l", "set log level")
 	cmd.Flags().Var(params.LogFormat, "log-format", "set log format")
 	cmd.Flags().StringVar(&params.LogTimestampFormat, "log-timestamp-format", "", "set log timestamp format (OPA_LOG_TIMESTAMP_FORMAT environment variable)")
+	cmd.Flags().BoolVarP(&params.StdIn, "std-in", "", false, "read input from os.Stdin rather than a static file; read timeout is 30 seconds")
```
Comment (srenatus) [inline cmd/exec.go:82]: Let's use the same flag as `opa eval` uses, please:
```
  -I, --stdin-input                                               read input document from stdin
```

File: cmd/internal/exec/std_in_reader.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	"time"
+)
+
+type stdInReader struct {
+	Reader  io.Reader
+	Timeout time.Duration
+}
+
+func (sr *stdInReader) ReadInput() (string, error) {
+	input := make(chan string, 1)
```
Comment (srenatus) [inline cmd/internal/exec/std_in_reader.go:17]: Who closes that channel? 🤔

Comment (colinjlacy) [reply cmd/internal/exec/std_in_reader.go:17]: Good catch

File: cmd/internal/exec/std_in_reader.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+				in.Scan()
+				line := in.Text()
+				if len(line) == 0 {
+					input <- strings.Join(lines, "\n")
+					return
+				}
+				lines = append(lines, line)
+			}
+		}
+	}(done)
```
Comment (srenatus) [inline cmd/internal/exec/std_in_reader.go:36]: 💭  I think the scoping rules would let you safely use `done` inside the `go func() { ... }()` body, wouldn't they?

File: cmd/internal/exec/std_in_reader.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+	for {
+		select {
+		case i := <-input:
+			return i, nil
+		case <-time.After(sr.Timeout):
+			done <- true
```
Comment (srenatus) [inline cmd/internal/exec/std_in_reader.go:43]: I think `close(done)` would be more slightly more idiomatic? Closed channels return their type's zero value; but we don't care for that value anyhow.

Comment (colinjlacy) [reply cmd/internal/exec/std_in_reader.go:43]: TIL 👍

**Comment by colinjlacy** [comment]
Hey folks, I'm stuck on one aspect of this and I could use some guidance.  In offline convos with @srenatus it was pointed out that trying to set a timeout on the goroutine that spawns the stdin reader creates a blocking thread that will continue to block until the main process exits.  That is to say that even with the timeout in the main goroutine, prompting it to continue and not wait for the child goroutine to send back data, the child will continue to block and wait for input from stdin.  That's not great as it's a known memory leak and unintentional block on further input (if further input is ever needed). So the timeout has some less-than ideal side effects.

On the other hand, as was discussed in the original issue, I'm concerned about the overall UX of adding functionality that could potentially hang indefinitely while the process awaits user input.  

In order to get something in, I'm more partial to the idea of preventing the leak/block and omitting the timeout, as it doesn't look like it's possible to get around blocking reads on stdin.  Any objections?

**Comment by srenatus** [comment]
I think it's OK to not have a timeout, and use the simplest possible thing here. It's also going to be on-par with the `opa eval -I`. 👍

File: cmd/exec_test.go [hunk 5]
```diff
@@ -1045,6 +1045,116 @@ func TestFailFlagCases(t *testing.T) {
...
+						t.Fatalf("unexpected error creating temp file: %q", err.Error())
+					}
+					if _, err := tempFile.Write([]byte(tt.input)); err != nil {
+						t.Fatalf("unexpeced error when writing to temp file: %q", err.Error())
+					}
+					if _, err := tempFile.Seek(0, 0); err != nil {
+						t.Fatalf("unexpected error when rewinding temp file: %q", err.Error())
+					}
+					oldStdin := os.Stdin
+					defer func() {
```
Comment (srenatus) [inline cmd/exec_test.go:1132]: [nit][ We could use `t.Cleanup()` here. But it doesn't make a difference, I believe.

File: docs/content/cli.md [hunk 6]
```diff
@@ -618,6 +618,7 @@ opa exec <path> [<path> [...]] [flags]
       --log-timestamp-format string          set log timestamp format (OPA_LOG_TIMESTAMP_FORMAT environment variable)
       --set stringArray                      override config values on the command line (use commas to specify multiple values)
       --set-file stringArray                 override config values with files on the command line (use commas to specify multiple values)
+      --std-in                               read input from `os.Stdin` rather than a static file
```
Comment (srenatus) [inline docs/content/cli.md:621]: This needs to be regenerated, I think?

**Review by srenatus — DISMISSED** [review]
Quite an overhaul of our tests there, but I think it's for the better. Thanks!

There's one nitpick that you can safely ignore (t.Cleanup) but the docs should probably be regenerated to match the flag.

👏 Thanks again!

**Comment by srenatus** [comment]
@johanfylling can you have a quick look and merge when it's convenient?

**Review by johanfylling — APPROVED** [review]
Thank you! 😃

**Pull request merged**